File: leet-code/108. Convert Sorted Array to Binary Search Tree/ConvertSortedArrayToBinarySearchTree108.java
package com.example;

import java.util.List;
import java.util.stream.IntStream;

public class ConvertSortedArrayToBinarySearchTree108 {
    public TreeNode sortedArrayToBST(int[] nums) {
        return createBST(IntStream.of(nums).boxed().toList());
    }

    private TreeNode createBST(List<Integer> list) {
        if (list.isEmpty()) return null;
        int center = (list.size() - 1) / 2;
        TreeNode node = new TreeNode(list.get(center));
        TreeNode leftChild = createBST(sliceList(list, 0, center));
        TreeNode rightChild = createBST(sliceList(list, center + 1, list.size() - center + 1));
        node.left = leftChild;
        node.right = rightChild;
        return node;
    }

    private List<Integer> sliceList(List<Integer> list, int skipNum, int limitNum) {
        return list.stream().skip(skipNum).limit(limitNum).toList();
    }
}

File: leet-code/108. Convert Sorted Array to Binary Search Tree/ConvertSortedArrayToBinarySearchTree108.kt
package com.example

class ConvertSortedArrayToBinarySearchTree108_2 {
    fun sortedArrayToBST(nums: IntArray): TreeNode? {
        fun sliceList(list: List<Int>, start: Int, end: Int): List<Int> {
            return list.slice(start..end).toMutableList()
        }

        fun createBST(nums: List<Int>): TreeNode? {
            if (nums.isEmpty()) return null
            val mid = (nums.size - 1) / 2
            val node = TreeNode(nums[mid])
            val leftChild = createBST(sliceList(nums, 0, mid - 1))
            val rightChild = createBST(sliceList(nums, mid + 1, nums.size - 1))
            node.left = leftChild
            node.right = rightChild
            return node
        }

        return createBST(nums.toList())
    }
}
